blog/2024-03-08-structuring-a-large-app.md

```diff
@@ -0,0 +1,391 @@
+
+## example_big_app/state.py
+
+_Common_ substates should be implemented in a separate module to avoid circular
+imports. This module should not import other modules in the app.
+
+## example_big_app/components
+
+This package contains reusable parts of the app, for example headers, footers,
+and menus. If a particular component requires state, the substate may be defined
+in the same module for locality. Any substate defined in a component module
+should only contain fields and event handlers pertaining to that individual
+component. 
+
+## example_big_app/template.py
+
+Most apps consist of a common page layout and structure which wraps the content
+for each page. It is helpful to define that layout in a separate module so it
+can be easily shared when constructing pages.
+
+In a Reflex app, any common frontend UI elements can be factored out as a
+function that returns a component. If a function accepts a function that
+returns a component, it can be used as a decorator as seen below.
+
+```python
+from typing import Callable
+
+import reflex as rx
+
+from .components.menu import menu
+from .components.navbar import navbar
+
+
+def template(page: Callable[[], rx.Component]) -> rx.Component:
+    return rx.vstack(
+        navbar(),
+        rx.hstack(
+            menu(),
+            rx.container(page()),
+        ),
+        width="100%",
+    )
+```
+
+## example_big_app/pages
+
+This package contains one module per page in the app. If a particular page depends on the state, the substate should
+be defined in the same module as the page. The page-returning function should be decorated with `rx.page()` and `template`.
+
+```python
+import reflex as rx
+
+from ..state import AuthState
+from ..template import template
+
+
+class LoginState(AuthState):
+    def handle_submit(self, form_data):
+        self.logged_in = authenticate(form_data["username"], form_data["password"])
+
+
+def login_field(name: str, **input_props):
+    return rx.hstack(
+        rx.text(name.capitalize()),
+        rx.input(name=name, **input_props),
+        width="100%",
+        justify="between",
+    )
+
+
+@rx.page(route="/login")
+@template
+def login():
+    return rx.card(
+        rx.form(
+            rx.vstack(
+                login_field("username"),
+                login_field("password", type="password"),
+                rx.button("Login"),
+                width="100%",
+                justify="center",
+            ),
+            on_submit=LoginState.handle_submit,
+        ),
+    )
+```
+
+## example_big_app/\__init\__.py
+
+The `__init__.py` file defines the top-level package. This is a great place to import all state, models, and
+pages that should be part of the app. Typically, components and helpers
+do not need to imported, because they will be imported by pages that use them (or they would be unused).
+
+```python
+from . import state, models
+from .pages import index, login, post, product, profile, schedule
```

Comment:
Are we recommending the decorator primarily now?

Reply:
decorator is really nice for delegating control over page title, on loads, etc to the page module without have to deal with circular imports. i would say it's recommended